File: tools/generate-portraits.py
#!/usr/bin/env python3
"""Charazay oyuncu portre havuzu — FAZ 17 kova tabanlı üretim + işleme boru hattı.

    python3 tools/generate-portraits.py <kova> <adet> [is_parcacigi]

    kova : akd | siyah | kuz | beyaz | afr | lat | asya
    adet : ZORUNLU — bu çağrıda kaç yeni portre üretileceği

KAYNAK VE LİSANS — DİKKAT
-------------------------
Görseller pollinations.ai üzerinden üretiliyor. Bu servisin TİCARİ KULLANIM LİSANSI
BELİRSİZDİR: Terms sayfası JavaScript gerektirdiği için okunamadı, kendi GitHub deposunda
lisans sorusu cevapsız duruyor. Kullanıcı bunu bilerek kabul etti — önce web sürümü,
Steam öncesinde gerekirse görseller değiştirilecek. Bu yüzden ÜRETİM adımı ayrı bir
fonksiyondadır (`uret_bir`); kaynak değişirse boru hattının geri kalanı (fon eşitleme,
kırpma, eleme, manifest) aynen kalır.

BORU HATTI
----------
üret → fon parlaklığını eşitle (hedef ~120, std ≤8) → 256x320'den 256x250'ye dar kırp
     → bozuk/bulanık/benzer olanı ele → manifest'i güncelle

Elenen dosyanın numarası ATLANMAZ: sıradaki üretim o numarayı doldurur.
Yeniden numaralama YAPILMAZ — oyuncular portre dosya adını kendi üzerlerinde saklar.

NOT: Bu makinede Python kurulu değilse aynı boru hattının Node karşılığı kullanılır:
    node tools/generate-portraits.js <kova> <adet>
"""
import json
import sys
import time
import urllib.error
import urllib.parse
import urllib.request
from pathlib import Path

try:
    from PIL import Image, ImageFilter, ImageStat
except ImportError:  # pragma: no cover
    print("Pillow gerekli:  pip install Pillow", file=sys.stderr)
    raise

ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / "assets" / "portraits"
MANIFEST = OUT / "manifest.json"

KOVALAR = ("akd", "siyah", "kuz", "beyaz", "afr", "lat", "asya")
BANTLAR = ("genc", "kidemli")
# Dağılım %45 genç / %55 kıdemli — oyuncu yaşı rand(18,36), 25 dahil genç.
GENC_PAY = 0.45

# ── Kilitlenmiş istem parçaları ───────────────────────────────────────────────────────
# Fon TEK: eski havuzda parlaklık 10,9–155,2 arasında geziniyordu (14 kat fark), kadro
# ekranı dağınık görünüyordu. Giysi TEK AİLE: kapüşonlu/tişört karışımı çıkarıldı.
FON = "neutral medium gray studio background"
GIYSI = "plain dark navy sleeveless basketball jersey"
KOVA_ETNIK = {
    "akd":   "Turkish Mediterranean man, olive skin, dark hair",
    "siyah": "African American man, dark skin",
    "kuz":   "Northern European man, fair skin, light brown hair",
    "beyaz": "white North American man, fair skin",
    "afr":   "West African man, very dark skin",
    "lat":   "Latin American man, tan skin, dark hair",
    "asya":  "East Asian man, straight black hair",
}
KOVA_BANT_YAS = {"genc": "age 19-25", "kidemli": "age 27-35"}
# Yüz çeşitliliği — etnik ifade ve fon SABİT kalır, yalnız bunlar döner.
LOOK = [
    "short hair, clean shaven",
    "buzz cut, short beard",
    "curly hair, athletic build",
    "shaved head, strong jawline",
    "fade haircut, goatee",
    "wavy hair, light stubble",
    "cropped hair, broad shoulders",
    "textured hair, light beard",
]

# ── Eleme eşikleri ────────────────────────────────────────────────────────────────────
HEDEF_PARLAKLIK = 120.0     # fon eşitleme hedefi
MIN_NETLIK = 90.0           # Laplace varyansı — altındaki bulanık sayılır
DHASH_MIN_MESAFE = 8        # bu Hamming mesafesinden yakın olan "aşırı benzer" sayılır
GENISLIK, YUKSEKLIK = 256, 320
KIRP_YUKSEKLIK = 230        # FAZ 17B: 256x320 → 256x230 (forma payı azalır, yüz büyür)


# ══ 1) ÜRETİM — kaynak değişirse yalnız burası değişir ════════════════════════════════
def istem(kova: str, bant: str, i: int) -> str:
    return (
        f"professional basketball player portrait headshot, {KOVA_ETNIK[kova]}, "
        f"{GIYSI}, {LOOK[i % len(LOOK)]}, {KOVA_BANT_YAS[bant]}, "
        f"{FON}, photorealistic, front facing, chest up, soft even lighting, "
        "single person, face fully visible, no text, no watermark, no logo"
    )


def uret_bir(kova: str, bant: str, i: int) -> bytes | None:
    """Tek portrenin ham baytlarını döndürür. Ağ/servis katmanı YALNIZ burada."""
    seed = abs(hash((kova, bant, i))) % 10_000_000
    q = urllib.parse.quote(istem(kova, bant, i))
    url = (
        f"https://image.pollinations.ai/prompt/{q}"
        f"?seed={seed}&width={GENISLIK}&height={YUKSEKLIK}&nologo=true"
    )
    req = urllib.request.Request(
        url,
        headers={"User-Agent": "CharazayPortraitBot/2.0", "Accept": "image/*"},
    )
    for deneme in range(6):
        try:
            with urllib.request.urlopen(req, timeout=120) as resp:
                data = resp.read()
            return data if len(data) >= 3000 else None
        except urllib.error.HTTPError as e:
            if e.code == 429:
                time.sleep(4 * (deneme + 1))
                continue
            return None
        except Exception:
            time.sleep(2 * (deneme + 1))
    return None


# ══ 2) İŞLEME — fon eşitleme + dar kırpma ═════════════════════════════════════════════
def fon_parlakligi(im: Image.Image) -> float:
    """Kenar şeritlerinin ortalama parlaklığı — özne değil FON ölçülür."""
    g = im.convert("L")
    w, h = g.size
    serit = [g.crop((0, 0, int(w * 0.12), h)), g.crop((int(w * 0.88), 0, w, h)),
             g.crop((0, 0, w, int(h * 0.10)))]
    toplam = sum(ImageStat.Stat(s).mean[0] * s.size[0] * s.size[1] for s in serit)
    alan = sum(s.size[0] * s.size[1] for s in serit)
    return toplam / max(1, alan)


def isle(im: Image.Image) -> Image.Image:
    """Fon parlaklığını HEDEF_PARLAKLIK'a çeker, sonra dar kırpar."""
    fark = HEDEF_PARLAKLIK - fon_parlakligi(im)
    if abs(fark) > 1.0:
        im = im.point(lambda v: max(0, min(255, int(v + fark))))
    w, h = im.size
    ust = int((h - KIRP_YUKSEKLIK) * 0.35)       # üstten biraz, altından çok kırp
    return im.crop((0, ust, w, ust + KIRP_YUKSEKLIK))


# ══ 3) ELEME ══════════════════════════════════════════════════════════════════════════
def netlik(im: Image.Image) -> float:
    g = im.convert("L").filter(ImageFilter.FIND_EDGES)
    return ImageStat.Stat(g).stddev[0] ** 2


def dhash(im: Image.Image) -> int:
    g = im.convert("L").resize((9, 8))
    px = list(g.getdata())
    bit = 0
    for y in range(8):
        for x in range(8):
            bit = (bit << 1) | int(px[y * 9 + x] < px[y * 9 + x + 1])
    return bit


def hamming(a: int, b: int) -> int:
    return bin(a ^ b).count("1")


def yuz_var_mi(im: Image.Image) -> bool:
    """Merkez bölgede ten rengi oranı — tek yüz vekili.

    NOT: gerçek yüz SAYISI tespiti OpenCV/dlib gerektirir. Bu fonksiyon onun yerine
    "merkezde makul büyüklükte tek bir ten yüzeyi var mı" sorusunu yanıtlar; boş,
    soyut ve özneyi çok uzakta bırakan kareleri eler.
    """
    w, h = im.size
    merkez = im.convert("RGB").crop((int(w * 0.28), int(h * 0.10), int(w * 0.72), int(h * 0.55)))
    px = list(merkez.getdata())
    ten = sum(
        1 for r, g, b in px
        if r > 60 and g > 35 and b > 20 and r > b and (max(r, g, b) - min(r, g, b)) > 12
    )
    oran = ten / max(1, len(px))
    return 0.30 <= oran <= 0.97


# ══ 4) MANİFEST ═══════════════════════════════════════════════════════════════════════
def manifest_oku() -> dict:
    try:
        return json.loads(MANIFEST.read_text(encoding="utf-8"))
    except Exception:
        return {"version": 2, "buckets": {}, "pattern": "%s_%s_%04d.jpg"}


def manifest_yaz() -> dict:
    """Diskteki GERÇEK dosyaları sayar — elle sayı girilmez."""
    m = {"version": 2, "buckets": {}, "pattern": "%s_%s_%04d.jpg"}
    for k in KOVALAR:
        m["buckets"][k] = {}
        for b in BANTLAR:
            n = 0
            while (OUT / f"{k}_{b}_{n:04d}.jpg").exists():
                n += 1
            m["buckets"][k][b] = n
    MANIFEST.write_text(json.dumps(m, indent=2, ensure_ascii=False) + "\n", encoding="utf-8")
    return m


def sonraki_sira(kova: str, bant: str) -> int:
    n = 0
    while (OUT / f"{kova}_{bant}_{n:04d}.jpg").exists():
        n += 1
    return n


# ══ 5) ANA AKIŞ ═══════════════════════════════════════════════════════════════════════
def mevcut_hashler(kova: str, bant: str) -> list[int]:
    out = []
    for p in sorted(OUT.glob(f"{kova}_{bant}_*.jpg")):
        try:
            out.append(dhash(Image.open(p)))
        except Exception:
            pass
    return out


def main() -> None:
    if len(sys.argv) < 3:
        print(__doc__)
        sys.exit(2)
    kova = sys.argv[1]
    if kova not in KOVALAR:
        print(f"bilinmeyen kova: {kova} (geçerli: {', '.join(KOVALAR)})", file=sys.stderr)
        sys.exit(2)
    adet = int(sys.argv[2])
    OUT.mkdir(parents=True, exist_ok=True)

    hashler = {b: mevcut_hashler(kova, b) for b in BANTLAR}
    yazilan = {b: 0 for b in BANTLAR}
    elenen = {"bozuk": 0, "bulanik": 0, "yuz": 0, "benzer": 0}

    i = 0
    denenen = 0
    while sum(yazilan.values()) < adet and denenen < adet * 4:
        denenen += 1
        # Bant seçimi DİSKTEKİ duruma bakar, sayaç turuna değil. Sabit tur küçük
        # partilerde bozuluyordu: bir kovaya 8 tane istendiğinde 8'i de "genc" oluyordu.
        # FAZ 17B §3: (a) bir kovada ikinci görsel daima diğer banda gider — hiçbir
        # bant 0'da kalmaz; (b) sonrası %45/%55 hedefinden geri kalan bandı doldurur.
        var_genc, var_kid = sonraki_sira(kova, "genc"), sonraki_sira(kova, "kidemli")
        toplam_v = var_genc + var_kid
        if toplam_v == 0 or var_genc == 0:
            bant = "genc"
        elif var_kid == 0:
            bant = "kidemli"
        else:
            bant = "genc" if var_genc / toplam_v < GENC_PAY else "kidemli"
        i += 1
        ham = uret_bir(kova, bant, sonraki_sira(kova, bant) + denenen)
        if not ham:
            elenen["bozuk"] += 1
            continue
        try:
            import io
            im = Image.open(io.BytesIO(ham)).convert("RGB")
        except Exception:
            elenen["bozuk"] += 1
            continue
        im = isle(im)
        if not yuz_var_mi(im):
            elenen["yuz"] += 1
            continue
        if netlik(im) < MIN_NETLIK:
            elenen["bulanik"] += 1
            continue
        h = dhash(im)
        if any(hamming(h, o) < DHASH_MIN_MESAFE for o in hashler[bant]):
            elenen["benzer"] += 1
            continue
        # Elenen dosyanın numarası atlanmaz: sıra her zaman ilk BOŞ numaradır.
        sira = sonraki_sira(kova, bant)
        im.save(OUT / f"{kova}_{bant}_{sira:04d}.jpg", "JPEG", quality=88, optimize=True)
        hashler[bant].append(h)
        yazilan[bant] += 1
        print(f"ok {kova}_{bant}_{sira:04d}.jpg")

    m = manifest_yaz()
    print(f"bitti: {sum(yazilan.values())}/{adet} yazıldı {yazilan} · elenen {elenen}")
    print(f"manifest: {json.dumps(m['buckets'][kova], ensure_ascii=False)}")


if __name__ == "__main__":
    main()

File: tools/generate-portraits.js
#!/usr/bin/env node
/* Charazay portre üretimi — FAZ 17 boru hattının Node karşılığı.
 *
 *     node tools/generate-portraits.js <kova> <adet> [--jobs=N]
 *
 * NEDEN İKİ BETİK VAR: tools/generate-portraits.py aynı boru hattının Python/Pillow
 * sürümüdür ve brifin istediği arayüzdür. Bu geliştirme makinesinde Python KURULU DEĞİL
 * ve node_modules içinde görüntü kütüphanesi (sharp/jimp/canvas) yok; bu yüzden işleme
 * adımları depoda zaten bulunan Playwright'ın headless Chromium'unda <canvas> ile
 * yapılıyor. İki betik aynı dosya adlarını, aynı eşikleri ve aynı manifest'i üretir.
 *
 * KAYNAK VE LİSANS — DİKKAT
 * Görseller pollinations.ai üzerinden üretiliyor; TİCARİ KULLANIM LİSANSI BELİRSİZ
 * (Terms sayfası JS gerektirdiği için okunamadı, GitHub'da lisans sorusu cevapsız).
 * Kullanıcı bunu bilerek kabul etti: önce web, Steam öncesinde gerekirse kaynak değişir.
 * Bu yüzden ÜRETİM adımı tek fonksiyonda durur (uretBir); kaynak değişirse fon eşitleme,
 * kırpma, eleme ve manifest aynen kalır.
 *
 * BORU HATTI: üret → fon parlaklığını eşitle (hedef ~120) → 256x320'den 256x250'ye dar
 * kırp → bozuk / bulanık / yüzsüz / aşırı benzer olanı ele → manifest'i güncelle.
 * Elenen dosyanın numarası ATLANMAZ; sıra her zaman ilk boş numaradır.
 */
const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..');
const OUT = path.join(ROOT, 'assets', 'portraits');
const MANIFEST = path.join(OUT, 'manifest.json');

const KOVALAR = ['akd', 'siyah', 'kuz', 'beyaz', 'afr', 'lat', 'asya'];
const BANTLAR = ['genc', 'kidemli'];
const GENC_PAY = 0.45;              /* %45 genç / %55 kıdemli */

/* Kilitlenmiş istem: fon TEK, giysi TEK AİLE. Eski havuzda fon parlaklığı 10,9-155,2
   arasında geziniyordu (14 kat) ve 8 farklı giysi vardı — kadro ekranı dağınıktı. */
const FON = 'neutral medium gray studio background';
const GIYSI = 'plain dark navy sleeveless basketball jersey';
const KOVA_ETNIK = {
  akd:   'Turkish Mediterranean man, olive skin, dark hair',
  siyah: 'African American man, dark skin',
  kuz:   'Northern European man, fair skin, light brown hair',
  beyaz: 'white North American man, fair skin',
  afr:   'West African man, very dark skin',
  lat:   'Latin American man, tan skin, dark hair',
  asya:  'East Asian man, straight black hair',
};
const BANT_YAS = { genc: 'age 19-25', kidemli: 'age 27-35' };
const LOOK = ['short hair, clean shaven', 'buzz cut, short beard', 'curly hair, athletic build',
  'shaved head, strong jawline', 'fade haircut, goatee', 'wavy hair, light stubble',
  'cropped hair, broad shoulders', 'textured hair, light beard'];

const HEDEF_PARLAKLIK = 120;
const MIN_NETLIK = 90;
const DHASH_MIN_MESAFE = 8;
const GENISLIK = 256, YUKSEKLIK = 320, KIRP_YUKSEKLIK = 250;

/* ══ 1) ÜRETİM — kaynak değişirse yalnız burası değişir ═════════════════════════════ */
function istem(kova, bant, i) {
  return `professional basketball player portrait headshot, ${KOVA_ETNIK[kova]}, ${GIYSI}, ` +
    `${LOOK[i % LOOK.length]}, ${BANT_YAS[bant]}, ${FON}, photorealistic, front facing, ` +
    'chest up, soft even lighting, single person, face fully visible, no text, no watermark, no logo';
}
function tohum(kova, bant, i) {
  let h = 5381;
  const s = kova + '|' + bant + '|' + i;
  for (let k = 0; k < s.length; k++) h = (((h << 5) + h) ^ s.charCodeAt(k)) >>> 0;
  return h % 10000000;
}
async function uretBir(kova, bant, i) {
  const url = 'https://image.pollinations.ai/prompt/' + encodeURIComponent(istem(kova, bant, i)) +
    `?seed=${tohum(kova, bant, i)}&width=${GENISLIK}&height=${YUKSEKLIK}&nologo=true`;
  for (let deneme = 0; deneme < 5; deneme++) {
    try {
      const r = await fetch(url, {
        headers: { 'User-Agent': 'CharazayPortraitBot/2.0', Accept: 'image/*' },
        signal: AbortSignal.timeout(150000),
      });
      if (r.status === 429) { await bekle(4000 * (deneme + 1)); continue; }
      if (!r.ok) { await bekle(1500 * (deneme + 1)); continue; }
      const buf = Buffer.from(await r.arrayBuffer());
      if (buf.length < 3000) return null;
      return buf;
    } catch (e) { await bekle(2000 * (deneme + 1)); }
  }
  return null;
}
const bekle = (ms) => new Promise(r => setTimeout(r, ms));

/* ══ 2-3) İŞLEME + ELEME — headless Chromium <canvas> ═══════════════════════════════ */
/* Tarayıcı içinde çalışır: fon parlaklığı ölçümü, normalize, dar kırpma, netlik
   (Laplace varyansı), ten-yüzeyi vekili ve dHash tek geçişte yapılır. */
const SAYFA_KODU = `
window.__isle = async function(dataUrl){
  const img = new Image();
  await new Promise((res, rej) => { img.onload = res; img.onerror = rej; img.src = dataUrl; });
  const W = ${GENISLIK}, H = ${YUKSEKLIK}, KH = ${KIRP_YUKSEKLIK};
  const c0 = document.createElement('canvas'); c0.width = W; c0.height = H;
  const x0 = c0.getContext('2d', { willReadFrequently: true });
  x0.drawImage(img, 0, 0, W, H);
  let d = x0.getImageData(0, 0, W, H);
  const px = d.data;
  const lum = (i) => 0.299*px[i] + 0.587*px[i+1] + 0.114*px[i+2];
  /* fon = sol/sağ kenar şeritleri + üst şerit (özne değil fon ölçülür) */
  let fonT = 0, fonN = 0;
  for (let y = 0; y < H; y++) for (let x = 0; x < W; x++) {
    const kenar = x < W*0.12 || x > W*0.88 || y < H*0.10;
    if (!kenar) continue;
    fonT += lum((y*W + x) * 4); fonN++;
  }
  const fon = fonT / Math.max(1, fonN);
  const fark = ${HEDEF_PARLAKLIK} - fon;
  if (Math.abs(fark) > 1) {
    for (let i = 0; i < px.length; i += 4) {
      px[i]   = Math.max(0, Math.min(255, px[i]   + fark));
      px[i+1] = Math.max(0, Math.min(255, px[i+1] + fark));
      px[i+2] = Math.max(0, Math.min(255, px[i+2] + fark));
    }
  }
  x0.putImageData(d, 0, 0);
  /* dar kırpma: üstten az, alttan çok — giysi payı azalır, yüz büyür */
  const ust = Math.round((H - KH) * 0.35);
  const c1 = document.createElement('canvas'); c1.width = W; c1.height = KH;
  const x1 = c1.getContext('2d', { willReadFrequently: true });
  x1.drawImage(c0, 0, ust, W, KH, 0, 0, W, KH);
  const k = x1.getImageData(0, 0, W, KH).data;
  const L = (i) => 0.299*k[i] + 0.587*k[i+1] + 0.114*k[i+2];
  /* fon ölçümü (kırpılmış hâlde, raporlama ve doğrulama için) */
  let fT = 0, fN = 0, kareT = 0;
  for (let y = 0; y < KH; y++) for (let x = 0; x < W; x++) {
    if (!(x < W*0.12 || x > W*0.88 || y < KH*0.10)) continue;
    const v = L((y*W + x) * 4); fT += v; kareT += v*v; fN++;
  }
  const fonOrt = fT / Math.max(1, fN);
  const fonStd = Math.sqrt(Math.max(0, kareT/Math.max(1,fN) - fonOrt*fonOrt));
  /* netlik: Laplace varyansı */
  let lT = 0, lK = 0, lN = 0;
  for (let y = 1; y < KH-1; y++) for (let x = 1; x < W-1; x++) {
    const i = (y*W + x) * 4;
    const v = 4*L(i) - L(i-4) - L(i+4) - L(i - W*4) - L(i + W*4);
    lT += v; lK += v*v; lN++;
  }
  const netlik = lK/lN - (lT/lN)*(lT/lN);
  /* ten yüzeyi vekili (gerçek yüz SAYISI tespiti OpenCV ister; bu onun yerine
     "merkezde makul büyüklükte tek ten yüzeyi var mı" sorusunu yanıtlar) */
  let ten = 0, tenN = 0;
  for (let y = Math.round(KH*0.10); y < KH*0.55; y++)
    for (let x = Math.round(W*0.28); x < W*0.72; x++) {
      const i = (y*W + x) * 4, r = k[i], g = k[i+1], b = k[i+2];
      tenN++;
      if (r > 60 && g > 35 && b > 20 && r > b && (Math.max(r,g,b) - Math.min(r,g,b)) > 12) ten++;
    }
  const tenOran = ten / Math.max(1, tenN);
  /* dHash 64 bit */
  const c2 = document.createElement('canvas'); c2.width = 9; c2.height = 8;
  const x2 = c2.getContext('2d', { willReadFrequently: true });
  x2.drawImage(c1, 0, 0, 9, 8);
  const s = x2.getImageData(0, 0, 9, 8).data;
  let hash = '';
  for (let y = 0; y < 8; y++) for (let x = 0; x < 8; x++) {
    const a = 0.299*s[(y*9+x)*4] + 0.587*s[(y*9+x)*4+1] + 0.114*s[(y*9+x)*4+2];
    const b = 0.299*s[(y*9+x+1)*4] + 0.587*s[(y*9+x+1)*4+1] + 0.114*s[(y*9+x+1)*4+2];
    hash += (a < b) ? '1' : '0';
  }
  return { jpeg: c1.toDataURL('image/jpeg', 0.88), fonOrt, fonStd, netlik, tenOran, hash };
};`;

function hamming(a, b) { let n = 0; for (let i = 0; i < a.length; i++) if (a[i] !== b[i]) n++; return n; }

/* ══ 4) MANİFEST — disk sayılır, elle sayı girilmez ═════════════════════════════════ */
function dosyaAdi(kova, bant, sira) {
  return `${kova}_${bant}_${String(sira).padStart(4, '0')}.jpg`;
}
function sonrakiSira(kova, bant) {
  let n = 0;
  while (fs.existsSync(path.join(OUT, dosyaAdi(kova, bant, n)))) n++;
  return n;
}
function manifestYaz() {
  const m = { version: 2, buckets: {}, pattern: '%s_%s_%04d.jpg' };
  for (const k of KOVALAR) { m.buckets[k] = {}; for (const b of BANTLAR) m.buckets[k][b] = sonrakiSira(k, b); }
  fs.writeFileSync(MANIFEST, JSON.stringify(m, null, 2) + '\n');
  return m;
}

/* ══ 5) ANA AKIŞ ════════════════════════════════════════════════════════════════════ */
async function main() {
  const kova = process.argv[2];
  const adet = parseInt(process.argv[3], 10);
  if (!KOVALAR.includes(kova) || !Number.isFinite(adet) || adet <= 0) {
    console.log('kullanım: node tools/generate-portraits.js <kova> <adet>');
    console.log('kovalar :', KOVALAR.join(' | '));
    process.exit(2);
  }
  fs.mkdirSync(OUT, { recursive: true });

  const { chromium } = require('playwright');
  const browser = await chromium.launch();
  const page = await browser.newPage();
  await page.addScriptTag({ content: SAYFA_KODU });

  /* mevcut dosyaların hash'leri — yeni parti eskilerin kopyası olmasın */
  const hashler = { genc: [], kidemli: [] };
  for (const b of BANTLAR) {
    for (let i = 0; i < sonrakiSira(kova, b); i++) {
      const p = path.join(OUT, dosyaAdi(kova, b, i));
      const dataUrl = 'data:image/jpeg;base64,' + fs.readFileSync(p).toString('base64');
      try { hashler[b].push((await page.evaluate(u => window.__isle(u), dataUrl)).hash); } catch (e) {}
    }
  }

  const yazilan = { genc: 0, kidemli: 0 };
  const elenen = { indirilemedi: 0, bozuk: 0, yuz: 0, bulanik: 0, benzer: 0 };
  const fonlar = [];
  let denenen = 0, i = 0;

  while (yazilan.genc + yazilan.kidemli < adet && denenen < adet * 4) {
    denenen++;
    const bant = (i % 20) < Math.round(GENC_PAY * 20) ? 'genc' : 'kidemli';
    i++;
    const ham = await uretBir(kova, bant, sonrakiSira(kova, bant) * 31 + denenen);
    if (!ham) { elenen.indirilemedi++; continue; }
    let r = null;
    try {
      r = await page.evaluate(u => window.__isle(u), 'data:image/jpeg;base64,' + ham.toString('base64'));
    } catch (e) { elenen.bozuk++; continue; }
    if (!r || !r.jpeg) { elenen.bozuk++; continue; }
    if (!(r.tenOran >= 0.30 && r.tenOran <= 0.97)) { elenen.yuz++; continue; }
    if (r.netlik < MIN_NETLIK) { elenen.bulanik++; continue; }
    if (hashler[bant].some(h => hamming(h, r.hash) < DHASH_MIN_MESAFE)) { elenen.benzer++; continue; }
    const sira = sonrakiSira(kova, bant);        /* elenen numara atlanmaz */
    fs.writeFileSync(path.join(OUT, dosyaAdi(kova, bant, sira)),
      Buffer.from(r.jpeg.split(',')[1], 'base64'));
    hashler[bant].push(r.hash);
    yazilan[bant]++;
    fonlar.push(r.fonOrt);
    process.stdout.write(`ok ${dosyaAdi(kova, bant, sira)}  fon=${r.fonOrt.toFixed(1)} netlik=${r.netlik.toFixed(0)}\n`);
  }

  await browser.close();
  const m = manifestYaz();
  const ort = fonlar.length ? fonlar.reduce((a, b) => a + b, 0) / fonlar.length : 0;
  const std = fonlar.length ? Math.sqrt(fonlar.reduce((a, b) => a + (b - ort) ** 2, 0) / fonlar.length) : 0;
  console.log(`\nbitti: ${yazilan.genc + yazilan.kidemli}/${adet} yazıldı ` +
    `(genc ${yazilan.genc}, kidemli ${yazilan.kidemli})`);
  console.log('elenen:', JSON.stringify(elenen));
  console.log(`fon parlaklığı: ort ${ort.toFixed(1)} · std ${std.toFixed(1)} (hedef ort ~120, std ≤8)`);
  console.log('manifest:', JSON.stringify(m.buckets[kova]));
}

main().catch(e => { console.error(e); process.exit(1); });
